<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tools for Assessing Clustering â€¢ ClustAssess</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Tools for Assessing Clustering">
<meta property="og:description" content="A set of tools for evaluating clustering similarity across methods 
    and method stability using element-centric clustering comparison (Gates et 
    al. (2019) &lt;doi:10.1038/s41598-019-44892-y&gt;). Additionally, this package 
    enables data-wide assessment of clustering robustness using proportion of
    ambiguously clustered pairs (Senbabaoglu et al. (2014) 
    &lt;doi:10.1038/srep06207&gt;), which can be used to infer the optimal number of 
    clusters in the data.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">ClustAssess</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/ClustAssess.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/comparing-soft-and-hierarchical.html">Comparing soft and hierarchical clusterings with element-centric similarity</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Core-Bioinformatics/ClustAssess/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="clustassess-tools-for-assessing-clustering" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#clustassess-tools-for-assessing-clustering" class="anchor"></a>ClustAssess: Tools for Assessing Clustering</h1></div>

<p>This repo contains the ClustAssess R package, which provides a set of tools for assessing clustering robustness.</p>
<ul>
<li><a href="#the-tools">The Tools</a></li>
<li><a href="#proportion-of-ambiguously-clustered-pairs-pac">Proportion of Ambiguously Clustered Pairs (PAC)</a></li>
<li><a href="#element-centric-clustering-similarity-ecs">Element Centric Clustering Similarity (ECS)</a></li>
<li><a href="#marker-gene-overlap">Marker gene overlap</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#references">References</a></li>
</ul>
</div>
<div id="the-tools" class="section level1">
<h1 class="hasAnchor">
<a href="#the-tools" class="anchor"></a>The Tools</h1>
<div id="proportion-of-ambiguously-clustered-pairs-pac" class="section level2">
<h2 class="hasAnchor">
<a href="#proportion-of-ambiguously-clustered-pairs-pac" class="anchor"></a>Proportion of Ambiguously Clustered Pairs (PAC)</h2>
<p>To assess clustering robustness, the proportion of ambiguously clustered pairs (PAC) [1] uses a consensus clustering. The rate of element co-clustering is recorded across various numbers of clusters, <em>k</em>. The lower the PAC, the stabler the clustering for that <em>k</em>.</p>
<p>ClustAssess uses a heirarchical clustering as base for the consensus clustering, and an optimized Rcpp [2] implementation to compute the PAC values. To calculate PAC, we write:</p>
<p><code>cc_res = consensus_cluster(your_data, n_reps=50, k_max=20, p_sample=0.8, p_feature=0.8)</code></p>
<p>It is important that the PAC has converged before using it to assess your data; the <code>pac_convergence</code> function can be used to visualize the PAC curves across iterations:</p>
<p><img src="https://raw.githubusercontent.com/Core-Bioinformatics/ClustAssess/main/docs/articles/ClustAssess_files/figure-html/pac-1.png" width="80%/"></p>
<p>As the curves have evened out, we surmize that PAC has converged in this case. If the PAC has not converged, increase the <code>n_reps</code> value.</p>
<p>If the dataset contains &gt;1000 elements, we recommend calculating a geometric sketch [3] of your data of size &lt;1000, and running PAC on that sketch.</p>
<p>A local minimum in the PAC landscape, as visualized below using the <code>pac_landscape</code> function, can be interpreted as an optimal <em>k</em> for the dataset:</p>
<p><img src="https://raw.githubusercontent.com/Core-Bioinformatics/ClustAssess/main/docs/reference/pac_landscape-1.png" width="80%/"></p>
</div>
<div id="element-centric-clustering-similarity-ecs" class="section level2">
<h2 class="hasAnchor">
<a href="#element-centric-clustering-similarity-ecs" class="anchor"></a>Element Centric Clustering Similarity (ECS)</h2>
<p>Element-centric similarity (ECS) [4] is a per-observation measure of clustering similarity. Briefly, ECS is obtained by constructing a cluster-induced element graph, and subsequently calculating the stationary probability distribution of a random walk with restarts (also known as a personalized Pagerank) over the elements.</p>
<p>In addition to providing per-element information on clustering agreement, ECS avoids several pitfalls associated with other measures of clustering similarity (refer to [4] for more detail).</p>
<p>To calculate ECS, the user first needs to create a <code>Clustering</code> object:</p>
<p><code>clustering_object1 = create_clustering(clustering_result1, alpha=0.9)</code></p>
<p>where <code>1-alpha</code> is the restart probability of the random walk. To compare two clusterings, we use the <code>element_sim_elscore</code> function:</p>
<p><code>ecs = element_sim_elscore(clustering_object1, clustering_object2)</code></p>
<p>and we can subsequently visualize the ECS on a PCA of the data:</p>
<p><img src="https://raw.githubusercontent.com/Core-Bioinformatics/ClustAssess/main/docs/articles/comparing-soft-and-hierarchical_files/figure-html/ecs-2.png" width="80%/"></p>
<p>In addition to flat disjoint clusterings (like the result of k-means for example), ClustAssess can also compare overlapping clusterings and hierarchical clusterings; refer to the <code>create_clustering</code> documentation for more detail.</p>
<p>Besides comparing two clusterings with <code>element_sim_elscore</code>, ClustAssess also enables computing the per-element consistency (aka frustration) between a set of clusterings with the <code>element_frustration</code> function, and comparing a set of clusterings with a ground truth partitioning of elements with the <code>element_agreement</code> function.</p>
</div>
<div id="marker-gene-overlap" class="section level2">
<h2 class="hasAnchor">
<a href="#marker-gene-overlap" class="anchor"></a>Marker Gene Overlap</h2>
<p>In the analysis of single-cell RNA-seq data, cluster marker genes are typically used to infer the cell type of clusters. However, two different clustering results may lead to two different sets of markers for the cells. The <code>marker_overlap</code> function allows the comparison of two sets of markers. The output can be either as the number of common marker genes per cell, or as Jaccard similarity (size of intersect divided by size of union) per cell.</p>
<p><img src="https://raw.githubusercontent.com/Core-Bioinformatics/ClustAssess/main/docs/articles/ClustAssess_files/figure-html/jsi-1.png" width="80%/"></p>
</div>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>ClustAssess can be installed from <a href="https://cran.r-project.org/index.html">CRAN</a>:</p>
<p><code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages("ClustAssess")</a></code></p>
<p>or from github using remotes:</p>
<p><code><a href="https://remotes.r-lib.org/reference/install_github.html">remotes::install_github("Core-Bioinformatics/ClustAssess")</a></code></p>
<p>The following packages are required for ClustAssess:</p>
<ul>
<li>ggplot2</li>
<li>dplyr</li>
<li>fastcluster</li>
<li>rlang</li>
<li>Matrix</li>
<li>igraph</li>
<li>magrittr</li>
<li>Rcpp</li>
<li>methods</li>
<li>stats</li>
</ul>
<p>To run all examples and vignettes, the following packages are also needed:</p>
<ul>
<li>knitr</li>
<li>rmarkdown</li>
<li>e1071</li>
<li>dbscan</li>
<li>dendextend</li>
<li>Seurat</li>
</ul>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>[1] È˜enbabaoÄŸlu, Y., Michailidis, G., &amp; Li, J. Z. (2014). Critical limitations of consensus clustering in class discovery. Scientific reports, 4(1), 1-13. <a href="https://doi.org/10.1038/srep06207" class="uri">https://doi.org/10.1038/srep06207</a></p>
<p>[2] Eddelbuettel, D., &amp; Balamuta, J. J. (2018). Extending R with C++: A brief introduction to Rcpp. The American Statistician, 72(1), 28-36. <a href="https://doi.org/10.1080/00031305.2017.1375990" class="uri">https://doi.org/10.1080/00031305.2017.1375990</a></p>
<p>[3] Hie, B., Cho, H., DeMeo, B., Bryson, B., &amp; Berger, B. (2019). Geometric sketching compactly summarizes the single-cell transcriptomic landscape. Cell systems, 8(6), 483-493. <a href="https://doi.org/10.1016/j.cels.2019.05.003" class="uri">https://doi.org/10.1016/j.cels.2019.05.003</a></p>
<p>[4] Gates, A. J., Wood, I. B., Hetrick, W. P., &amp; Ahn, Y. Y. (2019). Element-centric clustering comparison unifies overlaps and hierarchy. Scientific reports, 9(1), 1-13. <a href="https://doi.org/10.1038/s41598-019-44892-y" class="uri">https://doi.org/10.1038/s41598-019-44892-y</a></p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=ClustAssess">https://â€‹cloud.r-project.org/â€‹package=ClustAssess</a>
</li>
<li>Browse source code at <br><a href="https://github.com/Core-Bioinformatics/ClustAssess/">https://â€‹github.com/â€‹Core-Bioinformatics/â€‹ClustAssess/â€‹</a>
</li>
<li>Report a bug at <br><a href="https://github.com/Core-Bioinformatics/ClustAssess/issues">https://â€‹github.com/â€‹Core-Bioinformatics/â€‹ClustAssess/â€‹issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Arash Shahsavari <br><small class="roles"> Author, maintainer </small>  </li>
<li>Irina Mohorianu <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/r-hub/cranlogs.app"><img src="https://cranlogs.r-pkg.org/badges/ClustAssess" alt="rstudio mirror downloads"></a></li>
<li><a href="https://github.com/r-hub/cranlogs.app"><img src="https://cranlogs.r-pkg.org/badges/grand-total/ClustAssess" alt="rstudio mirror downloads"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Arash Shahsavari, Irina Mohorianu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
