<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Evaluating single-cell clustering with ClustAssess â€¢ ClustAssess</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Evaluating single-cell clustering with ClustAssess">
<meta property="og:description" content="ClustAssess">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ClustAssess</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ClustAssess.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/comparing-soft-and-hierarchical.html">Comparing soft and hierarchical clusterings with element-centric similarity</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Core-Bioinformatics/ClustAssess/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Evaluating single-cell clustering with ClustAssess</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Core-Bioinformatics/ClustAssess/blob/master/vignettes/ClustAssess.Rmd"><code>vignettes/ClustAssess.Rmd</code></a></small>
      <div class="hidden name"><code>ClustAssess.Rmd</code></div>

    </div>

    
    
<p>In this vignette we will illustrate several of the tools available in ClustAssess on a small single-cell RNA-seq dataset.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Core-Bioinformatics/ClustAssess">ClustAssess</a></span><span class="op">)</span>

<span class="co"># we will use the pbmc_small single-cell RNA-seq dataset available via Seurat</span>
<span class="fu"><a href="http://www.satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, group.by<span class="op">=</span><span class="st">'letter.idents'</span><span class="op">)</span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/setup-1.png" width="672"></p>
<div id="proportion-of-ambiguously-clustered-pairs" class="section level2">
<h2 class="hasAnchor">
<a href="#proportion-of-ambiguously-clustered-pairs" class="anchor"></a>Proportion of ambiguously clustered pairs</h2>
<p>The proportion of ambiguously clustered pairs (PAC) uses consensus clustering to infer the optimal number of clusters for the data, by observing how variably pairs of elements cluster together. The lower the PAC, the more stable the clustering. PAC was originally presented in <a href="https://doi.org/10.1038/srep06207" class="uri">https://doi.org/10.1038/srep06207</a>.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># retrieve scaled data for PAC calculation</span>
<span class="va">pbmc.data</span> <span class="op">=</span> <span class="fu"><a href="http://www.satijalab.org/seurat/reference/GetAssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">pbmc_small</span>, assay<span class="op">=</span><span class="st">'RNA'</span>, slot<span class="op">=</span><span class="st">'scale.data'</span><span class="op">)</span>

<span class="co"># perform consensus clustering</span>
<span class="va">cc.res</span> <span class="op">=</span> <span class="fu"><a href="../reference/consensus_cluster.html">consensus_cluster</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">pbmc.data</span><span class="op">)</span>, 
                           k_max<span class="op">=</span><span class="fl">30</span>, 
                           n_reps<span class="op">=</span><span class="fl">100</span>,
                           p_sample<span class="op">=</span><span class="fl">0.8</span>,
                           p_feature<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span>

<span class="co"># assess the PAC convergence for a few values of k - each curve should </span>
<span class="co"># have converged to some value</span>
<span class="va">k.plot</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">6</span>,<span class="fl">8</span>,<span class="fl">10</span><span class="op">)</span>
<span class="fu"><a href="../reference/pac_convergence.html">pac_convergence</a></span><span class="op">(</span><span class="va">cc.res</span>, <span class="va">k.plot</span><span class="op">)</span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/pac-1.png" width="672"></p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># visualize the final PAC across k - there seems to be a local maximum at k=7, </span>
<span class="co"># indicating that 7 clusters leads to a less stable clustering of the data than </span>
<span class="co"># nearby values of k</span>
<span class="fu"><a href="../reference/pac_landscape.html">pac_landscape</a></span><span class="op">(</span><span class="va">cc.res</span><span class="op">)</span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/pac-2.png" width="672"></p>
</div>
<div id="element-centric-clustering-comparison" class="section level2">
<h2 class="hasAnchor">
<a href="#element-centric-clustering-comparison" class="anchor"></a>Element-centric clustering comparison</h2>
<p>We compare the similarity of clustering results between methods (in this case, between Louvain community detection and k-means) using element-centric similarity, which quantifies the clustering similarity per cell. Higher ECS indicates that an observation is clustered similarly across methods. ECS was introduced in <a href="https://doi.org/10.1038/s41598-019-44892-y" class="uri">https://doi.org/10.1038/s41598-019-44892-y</a>.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># first, cluster with Louvain algorithm</span>
<span class="va">pbmc_small</span> <span class="op">=</span> <span class="fu"><a href="http://www.satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">pbmc_small</span>, resolution<span class="op">=</span><span class="fl">0.8</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Warning: Adding a command log without an assay associated with it</span>
<span class="va">louvain.clustering</span> <span class="op">=</span> <span class="fu"><a href="../reference/create_clustering.html">create_clustering</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span>
<span class="fu"><a href="http://www.satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, group.by<span class="op">=</span><span class="st">'seurat_clusters'</span><span class="op">)</span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/ecs-1.png" width="672"></p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># also cluster with PCA+k-means</span>
<span class="va">pbmc_pca</span> <span class="op">=</span> <span class="fu"><a href="http://www.satijalab.org/seurat/reference/Embeddings.html">Embeddings</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="st">'pca'</span><span class="op">)</span>
<span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">kmeans_clusters</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span><span class="op">(</span><span class="va">pbmc_pca</span>, 
                                              centers<span class="op">=</span><span class="fl">3</span>, 
                                              nstart<span class="op">=</span><span class="fl">10</span>, 
                                              iter.max<span class="op">=</span><span class="fl">1e3</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span>
<span class="va">kmeans.clustering</span> <span class="op">=</span> <span class="fu"><a href="../reference/create_clustering.html">create_clustering</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">kmeans_clusters</span><span class="op">)</span>
<span class="fu"><a href="http://www.satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, group.by<span class="op">=</span><span class="st">'kmeans_clusters'</span><span class="op">)</span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/ecs-2.png" width="672"></p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># where are the clustering results more similar?</span>
<span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">ecs</span> <span class="op">=</span> <span class="fu"><a href="../reference/element_sim_elscore.html">element_sim_elscore</a></span><span class="op">(</span><span class="va">louvain.clustering</span>, 
                                               <span class="va">kmeans.clustering</span><span class="op">)</span>
<span class="fu"><a href="http://www.satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="st">'ecs'</span><span class="op">)</span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/ecs-3.png" width="672"></p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">ecs</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.6632927</span></code></pre></div>
</div>
<div id="jaccard-similarity-of-cluster-markers" class="section level2">
<h2 class="hasAnchor">
<a href="#jaccard-similarity-of-cluster-markers" class="anchor"></a>Jaccard similarity of cluster markers</h2>
<p>As a common step in computational single-cell RNA-seq analyses, discriminative marker genes are identified for each cluster. These markers are then used to infer the cell type of the respective cluster. Here, we compare the markers obtained for each clustering method to ask: how similarly would each cell be interpreted across clustering methods? We compare the markers per cell using the Jaccard similarity (defined as the size of the intersect divided by the size of the overlap) of the marker gene lists. The higher the JSI, the more similar are the marker genes for that cell.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># first, we calculate the markers on the Louvain clustering</span>
<span class="fu"><a href="http://www.satijalab.org/seurat/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span> <span class="op">=</span> <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">seurat_clusters</span>
<span class="va">louvain.markers</span> <span class="op">=</span> <span class="fu"><a href="http://www.satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="va">pbmc_small</span>, 
                                 logfc.threshold<span class="op">=</span><span class="fl">1</span>, 
                                 min.pct<span class="op">=</span><span class="fl">0.0</span>, 
                                 test.use<span class="op">=</span><span class="st">'roc'</span>, 
                                 verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># then we get the markers on the k-means clustering</span>
<span class="fu"><a href="http://www.satijalab.org/seurat/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span> <span class="op">=</span> <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">kmeans_clusters</span>
<span class="va">kmeans.markers</span> <span class="op">=</span> <span class="fu"><a href="http://www.satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="va">pbmc_small</span>, 
                                logfc.threshold<span class="op">=</span><span class="fl">1</span>, 
                                min.pct<span class="op">=</span><span class="fl">0.0</span>, 
                                test.use<span class="op">=</span><span class="st">'roc'</span>, 
                                verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># next, compare the top 10 markers per cell</span>
<span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">marker.gene.jsi</span> <span class="op">=</span> <span class="fu"><a href="../reference/marker_overlap.html">marker_overlap</a></span><span class="op">(</span><span class="va">louvain.markers</span>, 
                                          <span class="va">kmeans.markers</span>, 
                                          <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">seurat_clusters</span>, 
                                          <span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">kmeans_clusters</span>, 
                                          n<span class="op">=</span><span class="fl">10</span>, 
                                          rank_by<span class="op">=</span><span class="st">'power'</span><span class="op">)</span>

<span class="co"># which cells have the same markers, regardless of clustering?</span>
<span class="fu"><a href="http://www.satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="st">'marker.gene.jsi'</span><span class="op">)</span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/jsi-1.png" width="672"></p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">marker.gene.jsi</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.6352273</span></code></pre></div>
</div>
<div id="element-wise-frustration" class="section level2">
<h2 class="hasAnchor">
<a href="#element-wise-frustration" class="anchor"></a>Element-wise frustration</h2>
<p>How consistently are cells clustered by k-means? We will rerun k-means clustering 20 times to investigate.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clustering.list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">n.reps</span> <span class="op">=</span> <span class="fl">20</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">n.reps</span><span class="op">)</span><span class="op">{</span>
  <span class="co"># we set nstart=1 and a fairly high iter.max - this should mean that</span>
  <span class="co"># the algorithm converges, and that the variability in final clusterings</span>
  <span class="co"># depends mainly on the random initial cluster assignments</span>
  <span class="va">km.res</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span><span class="op">(</span><span class="va">pbmc_pca</span>, centers<span class="op">=</span><span class="fl">3</span>, nstart<span class="op">=</span><span class="fl">1</span>, iter.max<span class="op">=</span><span class="fl">1e3</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span>
  <span class="va">clustering.list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="../reference/create_clustering.html">create_clustering</a></span><span class="op">(</span><span class="va">km.res</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># now, we calculate the element-wise frustration (ie consistency), which</span>
<span class="co"># performs pair-wise comparisons between all 20 clusterings; the </span>
<span class="co"># frustration is the average per-cell ECS across all comparisons. The higher</span>
<span class="co"># the frustration, the more consistently is that cell clustered across</span>
<span class="co"># random seeds.</span>
<span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">frustration</span> <span class="op">=</span> <span class="fu"><a href="../reference/element_frustration.html">element_frustration</a></span><span class="op">(</span><span class="va">clustering.list</span><span class="op">)</span>

<span class="co"># which cells are clustered more consistently?</span>
<span class="fu"><a href="http://www.satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="st">'frustration'</span><span class="op">)</span></code></pre></div>
<p><img src="ClustAssess_files/figure-html/frust-1.png" width="672"></p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">frustration</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.6712922</span></code></pre></div>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.0.2 (2020-06-22)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.7</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] ClustAssess_0.2.0 Seurat_3.1.4     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] TH.data_1.0-10      Rtsne_0.15          colorspace_1.4-1   </span>
<span class="co">#&gt;   [4] ellipsis_0.3.1      ggridges_0.5.3      rprojroot_1.3-2    </span>
<span class="co">#&gt;   [7] fs_1.4.2            rstudioapi_0.11     farver_2.0.3       </span>
<span class="co">#&gt;  [10] leiden_0.3.7        listenv_0.8.0       ggrepel_0.9.1      </span>
<span class="co">#&gt;  [13] mvtnorm_1.1-1       mathjaxr_1.4-0      codetools_0.2-16   </span>
<span class="co">#&gt;  [16] splines_4.0.2       mnormt_2.0.2        knitr_1.29         </span>
<span class="co">#&gt;  [19] TFisher_0.2.0       jsonlite_1.7.1      ica_1.0-2          </span>
<span class="co">#&gt;  [22] cluster_2.1.0       png_0.1-7           uwot_0.1.10        </span>
<span class="co">#&gt;  [25] sctransform_0.3.2   compiler_4.0.2      httr_1.4.2         </span>
<span class="co">#&gt;  [28] backports_1.1.8     assertthat_0.2.1    Matrix_1.2-18      </span>
<span class="co">#&gt;  [31] lazyeval_0.2.2      htmltools_0.5.1.1   tools_4.0.2        </span>
<span class="co">#&gt;  [34] rsvd_1.0.3          igraph_1.2.5        gtable_0.3.0       </span>
<span class="co">#&gt;  [37] glue_1.4.1          reshape2_1.4.4      RANN_2.6.1         </span>
<span class="co">#&gt;  [40] dplyr_1.0.0         Rcpp_1.0.5          Biobase_2.50.0     </span>
<span class="co">#&gt;  [43] pkgdown_1.6.1       vctrs_0.3.1         multtest_2.46.0    </span>
<span class="co">#&gt;  [46] ape_5.4-1           nlme_3.1-148        lmtest_0.9-38      </span>
<span class="co">#&gt;  [49] fastcluster_1.1.25  xfun_0.15           stringr_1.4.0      </span>
<span class="co">#&gt;  [52] globals_0.14.0      rbibutils_2.0       lifecycle_0.2.0    </span>
<span class="co">#&gt;  [55] irlba_2.3.3         future_1.21.0       MASS_7.3-51.6      </span>
<span class="co">#&gt;  [58] zoo_1.8-9           scales_1.1.1        ragg_1.1.2         </span>
<span class="co">#&gt;  [61] parallel_4.0.2      sandwich_3.0-0      RColorBrewer_1.1-2 </span>
<span class="co">#&gt;  [64] yaml_2.2.1          gridExtra_2.3       memoise_1.1.0      </span>
<span class="co">#&gt;  [67] reticulate_1.18     pbapply_1.4-3       ggplot2_3.3.2      </span>
<span class="co">#&gt;  [70] stringi_1.4.6       mutoss_0.1-12       desc_1.2.0         </span>
<span class="co">#&gt;  [73] plotrix_3.8-1       BiocGenerics_0.34.0 matrixStats_0.58.0 </span>
<span class="co">#&gt;  [76] Rdpack_2.1.1        rlang_0.4.10        pkgconfig_2.0.3    </span>
<span class="co">#&gt;  [79] systemfonts_1.0.1   evaluate_0.14       lattice_0.20-41    </span>
<span class="co">#&gt;  [82] ROCR_1.0-11         purrr_0.3.4         labeling_0.3       </span>
<span class="co">#&gt;  [85] patchwork_1.0.1     htmlwidgets_1.5.1   cowplot_1.1.1      </span>
<span class="co">#&gt;  [88] tidyselect_1.1.0    parallelly_1.24.0   RcppAnnoy_0.0.18   </span>
<span class="co">#&gt;  [91] plyr_1.8.6          magrittr_1.5        R6_2.4.1           </span>
<span class="co">#&gt;  [94] generics_0.0.2      multcomp_1.4-16     pillar_1.4.6       </span>
<span class="co">#&gt;  [97] sn_1.6-2            fitdistrplus_1.1-3  survival_3.1-12    </span>
<span class="co">#&gt; [100] tsne_0.1-3          tibble_3.0.3        future.apply_1.7.0 </span>
<span class="co">#&gt; [103] crayon_1.3.4        KernSmooth_2.23-17  tmvnsim_1.0-2      </span>
<span class="co">#&gt; [106] plotly_4.9.2.1      rmarkdown_2.3       grid_4.0.2         </span>
<span class="co">#&gt; [109] data.table_1.13.0   metap_1.4           digest_0.6.25      </span>
<span class="co">#&gt; [112] tidyr_1.1.0         numDeriv_2016.8-1.1 textshaping_0.3.3  </span>
<span class="co">#&gt; [115] stats4_4.0.2        munsell_0.5.0       viridisLite_0.3.0</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Arash Shahsavari, Irina Mohorianu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
