<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Stability-based parameter assessment • ClustAssess</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Stability-based parameter assessment">
<meta property="og:description" content="ClustAssess">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ClustAssess</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/ClustAssess.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/comparing-soft-and-hierarchical.html">Comparing soft and hierarchical clusterings with element-centric similarity</a>
    </li>
    <li>
      <a href="../articles/stability-based-parameter-assessment.html">Stability-based parameter assessment</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Core-Bioinformatics/ClustAssess/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Stability-based parameter assessment</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Core-Bioinformatics/ClustAssess/blob/HEAD/vignettes/stability-based-parameter-assessment.Rmd" class="external-link"><code>vignettes/stability-based-parameter-assessment.Rmd</code></a></small>
      <div class="hidden name"><code>stability-based-parameter-assessment.Rmd</code></div>

    </div>

    
    
<p>In this vignette we illustrate a data-driven pipeline for the assessment of optimal clustering parameters for a single-cell RNA-seq dataset. Popular pipelines for single-cell clustering (including <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>, <a href="https://scanpy.readthedocs.io/en/stable/" class="external-link">SCANPY</a>, and <a href="https://cole-trapnell-lab.github.io/monocle3/" class="external-link">Monocle v3</a>) consist of a sequence of steps, including normalization, dimensionality reduction, nearest-neighbor graph construction, and community detection on that graph. The obtained communities are then the clusters, which can be used for downstream analysis by looking at marker genes, trajectories, and more. ClustAssess offers a set of tools to assess the stability of parameters in the clustering pipeline, for the steps of dimensionality reduction, nearest-neighbor graph construction, and community detection.</p>
<p>To illustrate these tools, we will use a subset of cells from the Cuomo et al 2020 (<a href="https://www.nature.com/articles/s41467-020-14457-z" class="external-link">Single-cell RNA-sequencing of differentiating iPS cells reveals dynamic genetic effects on gene expression</a>) dataset. Specifically, 6 donors and 4 time points were chosen, comprising 1880 cells and 35430 expressed genes. The input is provided as a 35430 x 1880 expression level matrix containing the raw quantification (in csv format).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meta.path</span> <span class="op">=</span> <span class="st">'http://bioinf.stemcells.cam.ac.uk:3838/clustassess_1gmg5bq/cuomo_metadata.csv'</span>
<span class="va">counts.path</span> <span class="op">=</span> <span class="st">'http://bioinf.stemcells.cam.ac.uk:3838/clustassess_1gmg5bq/all_genes_with_RP_MT.csv.gz'</span>
<span class="va">cell_metadata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">meta.path</span>, sep<span class="op">=</span><span class="st">','</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">cell_metadata</span><span class="op">$</span><span class="va">donor</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">cell_metadata</span><span class="op">$</span><span class="va">donor</span><span class="op">)</span>
<span class="va">cell_metadata</span><span class="op">$</span><span class="va">day</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">cell_metadata</span><span class="op">$</span><span class="va">day</span><span class="op">)</span>

<span class="va">cuomo.counts</span> <span class="op">=</span> <span class="fu">read_csv</span><span class="op">(</span><span class="va">counts.path</span><span class="op">)</span>
<span class="co">#&gt; Registered S3 method overwritten by 'cli':</span>
<span class="co">#&gt;   method     from         </span>
<span class="co">#&gt;   print.boxx spatstat.geom</span>
<span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">35430</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">1881</span></span>
<span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span>
<span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> ","</span>
<span class="co">#&gt; <span style="color: #BB0000;">chr</span>    (1): gene_name</span>
<span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (1880): X21241_6_14, X21241_6_39, X21241_6_41, X21241_6_47, X21241_6_52,...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use <span style="color: #262626; background-color: #DADADA;">`spec()`</span> to retrieve the full column specification for this data.</span>
<span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set <span style="color: #262626; background-color: #DADADA;">`show_col_types = FALSE`</span> to quiet this message.</span>
<span class="va">gene_names</span> <span class="op">=</span> <span class="va">cuomo.counts</span><span class="op">$</span><span class="va">gene_name</span>
<span class="va">cuomo.counts</span> <span class="op">=</span> <span class="va">cuomo.counts</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">1880</span><span class="op">]</span>

<span class="va">cuomo</span> <span class="op">=</span>  <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">cuomo.counts</span>, row.names <span class="op">=</span> <span class="va">gene_names</span><span class="op">)</span>

<span class="va">cuomo</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">cuomo</span>, pattern <span class="op">=</span> <span class="st">'^MT-'</span>, col.name <span class="op">=</span> <span class="st">'percent.mito'</span><span class="op">)</span>
<span class="va">cuomo</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">cuomo</span>, pattern <span class="op">=</span> <span class="st">"^RP[SL][[:digit:]]"</span>, col.name <span class="op">=</span> <span class="st">'percent.rp'</span><span class="op">)</span>
<span class="va">one_level</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"cuomo"</span>, <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">cuomo</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>

<span class="va">cuomo</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">level.ident</span> <span class="op">=</span> <span class="va">one_level</span>
<span class="va">cuomo</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">day.ident</span> <span class="op">=</span> <span class="va">cell_metadata</span><span class="op">$</span><span class="va">day</span>
<span class="va">cuomo</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">donor.ident</span> <span class="op">=</span> <span class="va">cell_metadata</span><span class="op">$</span><span class="va">donor</span>
<span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">cuomo</span><span class="op">)</span> <span class="op">=</span> <span class="st">"donor.ident"</span>

<span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">cuomo.counts</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/gc.html" class="external-link">gc</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt;            used  (Mb) gc trigger   (Mb)  max used   (Mb)</span>
<span class="co">#&gt; Ncells  2844453 152.0    5489435  293.2   4862686  259.7</span>
<span class="co">#&gt; Vcells 29432554 224.6  134235283 1024.2 162914351 1243.0</span></code></pre></div>
<p>The analysis is based on the Seurat pipeline (<a href="https://doi.org/10.1016/j.cell.2021.04.048" class="external-link">Hao et al 2021, Integrated analysis of multimodal single-cell data</a>); after creating the object and recording the mitochondrial and ribosomal gene percentages, those genes are excluded from the downstream analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># remove MT and RP genes</span>
<span class="va">all.index</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">cuomo</span><span class="op">)</span>
<span class="va">MT.index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"^MT-"</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cuomo</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">RP.index</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"^RP[SL][[:digit:]]"</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cuomo</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">cuomo</span> <span class="op">=</span> <span class="va">cuomo</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="op">(</span><span class="va">all.index</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">MT.index</span><span class="op">)</span> <span class="op">|</span> <span class="op">(</span><span class="va">all.index</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">RP.index</span><span class="op">)</span>  <span class="op">)</span>, <span class="op">]</span></code></pre></div>
<p>The data is normalized and scaled using the Seurat’s <code>NormalizeData</code> and <code>ScaleData</code> methods. The variable feature set is obtained by running the <code>FindVariableFeatures</code> function from the same package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cuomo</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">cuomo</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="va">cuomo</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">cuomo</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">3000</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>

<span class="va">features</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dimnames.html" class="external-link">dimnames</a></span><span class="op">(</span><span class="va">cuomo</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="va">var_features</span> <span class="op">=</span> <span class="va">cuomo</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">var.features</span>
<span class="va">n_abundant</span> <span class="op">=</span> <span class="fl">3000</span>
<span class="va">most_abundant_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cuomo</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">cuomo</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">)</span>,
                                                       decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span>

<span class="va">cuomo</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">cuomo</span>, features <span class="op">=</span> <span class="va">features</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>

<span class="va">cuomo</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">cuomo</span>,
               npcs <span class="op">=</span> <span class="fl">30</span>,
               approx <span class="op">=</span> <span class="cn">F</span>,
               verbose <span class="op">=</span> <span class="cn">F</span>,
               features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">most_abundant_genes</span>, <span class="va">cuomo</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">var.features</span><span class="op">)</span><span class="op">)</span>
<span class="va">cuomo</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">cuomo</span>,reduction <span class="op">=</span> <span class="st">"pca"</span>,
                dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,
                n.neighbors <span class="op">=</span> <span class="fl">30</span>,
                min.dist <span class="op">=</span> <span class="fl">0.3</span>,
                metric <span class="op">=</span> <span class="st">"cosine"</span>,
                verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="va">raw_umap</span> <span class="op">=</span> <span class="va">cuomo</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">umap</span><span class="op">@</span><span class="va">cell.embeddings</span></code></pre></div>
<div class="section level1">
<h1 id="quality-checks">Quality checks<a class="anchor" aria-label="anchor" href="#quality-checks"></a>
</h1>
<p>QC overview for the Cuomo dataset.</p>
<p>A time point induced separation of cells; B distributions of cells per donor, we notice a good overlap across all 6 donors, supporting the assumption of no significant batch effect. C – F distributions of sequencing depths (C), number of features (D), and percentages of reads incident to mitochondrial genes (E) and ribosomal genes (F) illustrated using the colour gradient.</p>
<p>No significant effects are noted.</p>
<p><img src="stability-based-parameter-assessment_files/figure-html/qc-1.png" width="768"></p>
<p><em>Note</em>: The following sections provide an example of applying the ClustAssess summary assessment methods. The values of the parameters should be varied and critically assessed for each individual dataset. Also, we recommend increasing the number of repetitions used for stability inference, to ensure reliable and robust conclusions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">n_repetitions</span> <span class="op">=</span> <span class="fl">30</span></code></pre></div>
<p>The following analysis is performed by varying the seed multiple times. The <code>n_cores</code> user-defined parameter indicates the number of parallel threads used to run the pipeline. Setting the value of 1 will result in a sequential run of all iterations. We encourage the usage of a higher number of cores when possible, as it will decrease the runtime of the assessment pipeline. Please note that the number of cores should be chosen carefully, as each newly created process will add to the memory usage.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">n_cores</span> <span class="op">=</span> <span class="fl">1</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="dimensionality-reduction">Dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"></a>
</h1>
<p>The first parameter that we will evaluate and note to substantially influence the dimensionality reduction (linear or non-linear) is the feature set. The Seurat pipeline relies on the highly variable genes (default) for the computation of the Principal Components. The top x most variable genes or most abundant genes can be selected, where x varies according to the characteristics of the dataset; frequently used values are 500, 1000, 2000, 3000 etc.</p>
<p>To showcase the importance of this parameter, we assess three different sets of features:</p>
<ul>
<li>
<strong>highly variable genes</strong>: this set can be obtained in the Seurat pipeline by using either the <code>SCTransform</code> or <code>FindVariableFeatures</code> method. Depending on the method, using the default parameters will get a number of 2000 or 3000 genes.</li>
<li>
<strong>most abundant genes</strong>: this set can be obtained by sorting the genes based on their expression level.</li>
<li>
<strong>intersection between highly variable and most abundant genes</strong>: this set is obtained by intersecting a subset of the most abundant genes and the whole highly variable feature set. Taking the intersection will lead to a smaller number of genes compared to the initial sets.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ma_hv_genes_intersection</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">most_abundant_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3000</span><span class="op">]</span>, <span class="va">var_features</span><span class="op">)</span>
<span class="va">steps</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">500</span>, to <span class="op">=</span> <span class="fl">3000</span>, by <span class="op">=</span> <span class="fl">500</span><span class="op">)</span>
<span class="va">ma_hv_steps</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">steps</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">most_abundant_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">x</span><span class="op">]</span>, <span class="va">var_features</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></code></pre></div>
<p><code>get_feature_stability</code> is a function that explores the stability of a feature set. The input consists of a normalized expression matrix, that is summarized using either PCA or UMAP. The obtained embedding is then processed using the Leiden community detection method (on default values as described in the Seurat pipeline). This process is iterated across random seeds. The method outputs a list containing an UMAP embedding, the cell assignment corresponding to the most frequent partition and the Element-Centric Consistency (ECC) across all communities obtained throughout the iterations. The higher the Element-Centric Consistency, the more stable the clustering.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pca_feature_stability_object</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_feature_stability.html">get_feature_stability</a></span><span class="op">(</span>data_matrix <span class="op">=</span> <span class="va">cuomo</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">scale.data</span>,
                                                       feature_set <span class="op">=</span> <span class="va">most_abundant_genes</span>,
                                                       steps <span class="op">=</span> <span class="va">steps</span>,
                                                       n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,
                                                       feature_type <span class="op">=</span> <span class="st">"MA"</span>,
                                                       graph_reduction_type <span class="op">=</span> <span class="st">"PCA"</span>,
                                                       npcs <span class="op">=</span> <span class="fl">30</span>,
                                                       min_dist <span class="op">=</span> <span class="fl">0.3</span>,
                                                       n_neighbors <span class="op">=</span> <span class="fl">30</span>,
                                                       metric <span class="op">=</span> <span class="st">"cosine"</span>,
                                                       ncores <span class="op">=</span> <span class="va">n_cores</span>,
                                                       ecs_thresh <span class="op">=</span> <span class="fl">1</span>,
                                                       algorithm <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
                                 <span class="fu"><a href="../reference/get_feature_stability.html">get_feature_stability</a></span><span class="op">(</span>data_matrix <span class="op">=</span> <span class="va">cuomo</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">scale.data</span>,
                                                       feature_set <span class="op">=</span> <span class="va">var_features</span>,
                                                       steps <span class="op">=</span> <span class="va">steps</span>,
                                                       n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,
                                                       feature_type <span class="op">=</span> <span class="st">"HV"</span>,
                                                       graph_reduction_type <span class="op">=</span> <span class="st">"PCA"</span>,
                                                       npcs <span class="op">=</span> <span class="fl">30</span>,
                                                       min_dist <span class="op">=</span> <span class="fl">0.3</span>,
                                                       n_neighbors <span class="op">=</span> <span class="fl">30</span>,
                                                       metric <span class="op">=</span> <span class="st">"cosine"</span>,
                                                       ncores <span class="op">=</span> <span class="va">n_cores</span>,
                                                       ecs_thresh <span class="op">=</span> <span class="fl">1</span>,
                                                       algorithm <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
                                 <span class="fu"><a href="../reference/get_feature_stability.html">get_feature_stability</a></span><span class="op">(</span>data_matrix <span class="op">=</span> <span class="va">cuomo</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">scale.data</span>,
                                                       feature_set <span class="op">=</span> <span class="va">ma_hv_genes_intersection</span>,
                                                       steps <span class="op">=</span> <span class="va">ma_hv_steps</span>,
                                                       n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,
                                                       feature_type <span class="op">=</span> <span class="st">"MA_HV"</span>,
                                                       graph_reduction_type <span class="op">=</span> <span class="st">"PCA"</span>,
                                                       npcs <span class="op">=</span> <span class="fl">30</span>,
                                                       min_dist <span class="op">=</span> <span class="fl">0.3</span>,
                                                       n_neighbors <span class="op">=</span> <span class="fl">30</span>,
                                                       metric <span class="op">=</span> <span class="st">"cosine"</span>,
                                                       ncores <span class="op">=</span> <span class="va">n_cores</span>,
                                                       ecs_thresh <span class="op">=</span> <span class="fl">1</span>,
                                                       algorithm <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The function <code>get_feature_stability</code> outputs the stability of different feature sets and concatenates the results to ease cross-comparisons. The plot illustrates the stability assessment, across 30 runs, each with different random seeds, on the three gene sets described above. We assessed the stability of each set on incremental number of selected genes. Thus, the boxplots are arranged in groups of three (the number of evaluated feature sets); above each boxplot we specify the number of elements per subset. For the Cuomo case study, a conclusion that is supported based on this plot is that the most abundant set is more stable than the other two options, since it has higher Element-Centric Consistency scores.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_feature_stability_boxplot.html">plot_feature_stability_boxplot</a></span><span class="op">(</span><span class="va">pca_feature_stability_object</span>, text_size  <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0</span><span class="op">)</span>,
        legend.justification <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="stability-based-parameter-assessment_files/figure-html/stab_boxplot-1.png" width="768"></p>
<p>Another angle for assessing the stability is centered on the comparison between consecutive steps, for each feature set, performed using the Element-Centric Similarity on the most frequent partitions from each step. The aim is evaluating the effect of increasing the number of genes on the final partitions, and, indirectly, determining the transition from the signal to the noise zone. For the Cuomo case study, we observe an increase in similarity between consecutive steps for the most abundant and highly variable genes, suggesting that selecting more genes would lead to a more robust partitioning.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_feature_stability_ecs_incremental.html">plot_feature_stability_ecs_incremental</a></span><span class="op">(</span><span class="va">pca_feature_stability_object</span>, dodge_width <span class="op">=</span> <span class="fl">1</span>, text_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0</span><span class="op">)</span>,
        legend.justification <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="stability-based-parameter-assessment_files/figure-html/stab_ecs_inc-1.png" width="768"></p>
<p>To enhance the summarized information presented as boxplots, we use <code>plot_feature_stability_mb_facet</code> to visualize the corresponding facet plot displaying the most frequent partitions for each feature set and step on an UMAP embedding. This plot reveals additional insights i.e. the stability of the most abundant genes might be explained in part by the scattered distribution of the cells across multiple islands. The highly variable and the intersection gene sets tend to produce more compact partitions, that may be linked to biologically relevant cell types.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_feature_stability_mb_facet.html">plot_feature_stability_mb_facet</a></span><span class="op">(</span><span class="va">pca_feature_stability_object</span>, text_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="stability-based-parameter-assessment_files/figure-html/stab_mb-1.png" width="768"></p>
<p>To further understand the areas of instability, we provide an additional plot where we display, on the same UMAPs as before, the Element-Centric Consistency score of the partitions obtained for each step and feature set. For the Highly Variable and the intersection sets, we note that the instability appears in groups of cells with high density. These visualizations can help us understand the relationship between stability and the topology of the data.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_feature_stability_ecs_facet.html">plot_feature_stability_ecs_facet</a></span><span class="op">(</span><span class="va">pca_feature_stability_object</span><span class="op">)</span></code></pre></div>
<p><img src="stability-based-parameter-assessment_files/figure-html/stab_ecs_fa-1.png" width="768"></p>
<p>For the Cuomo case study, based on the plots presented above, we conclude that the most abundant gene set is not a suitable choice, due to the presence of multiple isolated groups of cells. Thus, we settle on using the highly variable genes set, as it produces a compact distribution of cells that is consistent across different seeds.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cuomo</span> <span class="op">=</span> <span class="va">cuomo</span><span class="op">[</span><span class="va">var_features</span>, <span class="op">]</span>

<span class="va">cuomo</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">cuomo</span>,
               npcs <span class="op">=</span> <span class="fl">30</span>,
               approx <span class="op">=</span> <span class="cn">F</span>,
               verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="va">cuomo</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">cuomo</span>,reduction <span class="op">=</span> <span class="st">"pca"</span>,
                dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,
                n.neighbors <span class="op">=</span> <span class="fl">30</span>,
                min.dist <span class="op">=</span> <span class="fl">0.3</span>,
                metric <span class="op">=</span> <span class="st">"cosine"</span>,
                verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="graph-construction">Graph construction<a class="anchor" aria-label="anchor" href="#graph-construction"></a>
</h1>
<p>The next step in a standard single-cell analysis pipeline is building the graph using the nearest neighbour algorithm. The following parameters influence the final partitioning:</p>
<ul>
<li>
<strong>base embedding</strong>: the graph can be built on either the PCA or the UMAP embedding (using the expression matrix isn’t recommended, as the distances would be noisier and the runtime would increase)</li>
<li><strong>the number of neighbours</strong></li>
<li>
<strong>the graph type</strong>: the graph can be either unweighted (NN case) or based on a weighted Shared-Nearest Neighbours (SNN) graph. For the latter, the weights are computed using the Jaccard Similarity Index (JSI) between the neighbourhoods of two cells.</li>
</ul>
<p><code>get_nn_conn_comps</code> is a method used to link the number of neighbours and the number of connected components (a connected component is a subgraph within which there exists a path between every pair of nodes). The method accepts as input both PCA and UMAP reductions; the distribution of the number of connected components obtainable for the given number of neighbours is calculated across random seeds. The output is a list containing, for each number of neighbours, an array with the number of connected components obtained on different seeds. As before, the objects can be concatenated to facilitate comparisons between different configurations and reductions.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nn_conn_comps_object</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_nn_conn_comps.html">get_nn_conn_comps</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cuomo</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span><span class="op">@</span><span class="va">cell.embeddings</span>,
                                           n_neigh_sequence <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">5</span>, to <span class="op">=</span> <span class="fl">30</span>, by <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span>,
                                           n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,
                                           graph_reduction_type <span class="op">=</span> <span class="st">"UMAP"</span>,
                                           ncores <span class="op">=</span> <span class="va">n_cores</span>,
                                           min_dist <span class="op">=</span> <span class="fl">0.3</span>,
                                           n_neighbors <span class="op">=</span> <span class="fl">30</span>,
                                           metric <span class="op">=</span> <span class="st">"cosine"</span><span class="op">)</span>,
                         <span class="fu"><a href="../reference/get_nn_conn_comps.html">get_nn_conn_comps</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cuomo</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">scale.data</span>,
                                           n_neigh_sequence <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">5</span>, to <span class="op">=</span> <span class="fl">30</span>, by <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span>,
                                           n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,
                                           graph_reduction_type <span class="op">=</span> <span class="st">"PCA"</span>,
                                           ncores <span class="op">=</span> <span class="va">n_cores</span>,
                                           nv <span class="op">=</span> <span class="fl">30</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The following plot describes the covariation between the number neighbours and the number of connected components obtained using both PCA and UMAP reductions as base for graph building. As the number of neighbours increases, the number of connected components decreases (this is an expected result, as increasing the number of neighbours result in a better connected graph). Please note that the number of connected components provides a lower bound on the number of clusters we can obtain by downstream community detection algorithms such as Louvain and Leiden.</p>
<p>Another comparison may be performed on the PCA and UMAP embeddings to assess graph connectivity. We note that, when using UMAPs, groups of cells are more separated compared to PCAs, for which a quick convergence to a connected graph (which is a graph where there exists a path between every two nodes) is obtained. The link between the number of connected components and the functional interpretation of the results is that the former could be interpreted as the minimum number of clusters that could be obtained with the given number of neighbours e.g. for the UMAP case, to obtain 5 communities from the clustering step, we would require at least 10 nearest neighbours.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_connected_comps_evolution.html">plot_connected_comps_evolution</a></span><span class="op">(</span><span class="va">nn_conn_comps_object</span><span class="op">)</span>
<span class="co">#&gt; Warning: Transformation introduced infinite values in continuous y-axis</span></code></pre></div>
<p><img src="stability-based-parameter-assessment_files/figure-html/comps_evo-1.png" width="768"></p>
<p><code>get_nn_importance</code> is a method used for assessing the impact of number of nearest neighbours, of the graph type (NN or SNN) and of the base embedding (PCA or UMAP) in determining the final number of clusters. The input is either a normalized expression matrix or a PCA embedding, a range for the number of neighbours and the reduction type used for building the graph. The graph type is not provided, as the method will check both cases i.e. NN and SNN. The output comprises a list of configurations (reduction type, and the graph type). Each configuration will contain a list of unique partitions obtained from multiple runs (iterations) for each selected value of the nearest neighbours.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nn_importance_object</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">c</span>,
                              <span class="fu"><a href="../reference/get_nn_importance.html">get_nn_importance</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cuomo</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">scale.data</span>,
                                         n_neigh_sequence <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">5</span>, to <span class="op">=</span> <span class="fl">30</span>, by <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,
                                         n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,
                                         graph_reduction_type <span class="op">=</span> <span class="st">"PCA"</span>,
                                         ecs_thresh <span class="op">=</span> <span class="fl">1</span>,
                                         ncores <span class="op">=</span> <span class="va">n_cores</span>,
                                         algorithm <span class="op">=</span> <span class="fl">1</span>,
                                         nv <span class="op">=</span> <span class="fl">30</span><span class="op">)</span>,
                              <span class="fu"><a href="../reference/get_nn_importance.html">get_nn_importance</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cuomo</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span><span class="op">@</span><span class="va">cell.embeddings</span>,
                                         n_neigh_sequence <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">5</span>, to <span class="op">=</span> <span class="fl">30</span>, by <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,
                                         n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,
                                         graph_reduction_type <span class="op">=</span> <span class="st">"UMAP"</span>,
                                         ecs_thresh <span class="op">=</span> <span class="fl">1</span>,
                                         ncores <span class="op">=</span> <span class="va">n_cores</span>,
                                         algorithm <span class="op">=</span> <span class="fl">1</span>,
                                         min_dist <span class="op">=</span> <span class="fl">0.3</span>,
                                         n_neighbors <span class="op">=</span> <span class="fl">30</span>,
                                         metric <span class="op">=</span> <span class="st">"cosine"</span><span class="op">)</span>,
                              SIMPLIFY <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
<p>We input the resulting object into <code>plot_n_neigh_k_correspondence</code>, to generate the following plot; it summarizes the relationship between the number of nearest neighbours and k, the number of clusters. As for the # neighbours vs # connected components correspondence plot, we note a similar descending trend of the number of clusters as the number of nearest neighbours increases. The plot also illustrates the difference between the two graph types: SNN has a tighter distribution of k (over multiple iterations) compared to NN. If the initial object contains graphs based on both UMAP and PCA embedding, this plot will showcase the impact of this choice, as well. For the Cuomo dataset, we notice that the UMAP-based graph is partitioned into a greater number of clusters compared to the PCA case.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_n_neigh_k_correspondence.html">plot_n_neigh_k_correspondence</a></span><span class="op">(</span><span class="va">nn_importance_object</span><span class="op">)</span></code></pre></div>
<p><img src="stability-based-parameter-assessment_files/figure-html/n_k_corr-1.png" width="768"></p>
<p>The stability of these parameters can be also evaluated using the Element-Centric Consistency applied on the partition list obtained over the multiple runs. The following summary plot underlines the evolution of the consistency as the number of neighbours increases; the PCA graphs yields stable results even for smaller number of nearest neighbours, whereas the UMAP exhibits a stability improvement as the number of neighbours increases. For the same number of nearest neighbors, SNN is more stable than NN.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_n_neigh_ecs.html">plot_n_neigh_ecs</a></span><span class="op">(</span><span class="va">nn_importance_object</span><span class="op">)</span></code></pre></div>
<p><img src="stability-based-parameter-assessment_files/figure-html/n_neigh_ecs-1.png" width="768"></p>
</div>
<div class="section level1">
<h1 id="graph-clustering">Graph clustering<a class="anchor" aria-label="anchor" href="#graph-clustering"></a>
</h1>
<p>The final step in a standard single-cell analysis pipeline is applying a graph-based clustering method. Choosing a community detection algorithm has a significant impact on the partitioning results. Using the <code>get_clustering_difference</code> we assess the stability and reproducibility of results obtained using various graph clustering methods available in the Seurat package: Louvain, Louvain refined, SLM and Leiden.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cuomo</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">cuomo</span>,
               npcs <span class="op">=</span> <span class="fl">30</span>,
               approx <span class="op">=</span> <span class="cn">F</span>,
               verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="va">cuomo</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">cuomo</span>,reduction <span class="op">=</span> <span class="st">"pca"</span>,
                dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,
                n.neighbors <span class="op">=</span> <span class="fl">30</span>,
                min.dist <span class="op">=</span> <span class="fl">0.3</span>,
                metric <span class="op">=</span> <span class="st">"cosine"</span>,
                verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="va">adj_matrix</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">cuomo</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">umap</span><span class="op">@</span><span class="va">cell.embeddings</span>, k.param <span class="op">=</span> <span class="fl">25</span>, nn.method <span class="op">=</span> <span class="st">"rann"</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span><span class="op">$</span><span class="va">snn</span>

<span class="va">start_time_clustering_method</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">clustering_diff_obj</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_clustering_difference.html">get_clustering_difference</a></span><span class="op">(</span>graph_adjacency_matrix <span class="op">=</span> <span class="va">adj_matrix</span>,
                                                resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0.5</span>, to <span class="op">=</span> <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,
                                                n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,
                                                ecs_thresh <span class="op">=</span> <span class="fl">1</span>,
                                                ncores <span class="op">=</span> <span class="va">n_cores</span>,
                                                algorithm <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>
<span class="co">#&gt; generating partitions for Louvain method</span>
<span class="co">#&gt; generating partitions for Louvain.refined method</span>
<span class="co">#&gt; generating partitions for SLM method</span>
<span class="co">#&gt; generating partitions for Leiden method</span></code></pre></div>
<p>The stability of the methods is evaluated using the Element-Centric Consistency (ECC), which is applied on the partition list obtained over multiple runs. The following summary plot underlines the variation of the consistency as the resolution parameter increases. Above each boxplot, we display the number of clusters corresponding to the most frequent partition for the clustering method with the specific resolution parameter. Increasing the resolution leads to more clusters; in addition, we note consistently high stability for all four algorithms, although Leiden seems to perform noticeably worse for the resolution value of 0.6.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_clustering_difference_boxplot.html">plot_clustering_difference_boxplot</a></span><span class="op">(</span><span class="va">clustering_diff_obj</span><span class="op">)</span></code></pre></div>
<p><img src="stability-based-parameter-assessment_files/figure-html/diff_boxplot-1.png" width="768"></p>
<p>To further underline the areas of instability, we provide an additional plot where we display, on the same UMAPs as before, the ECC score of the partitions obtained for each resolution parameter with each clustering method. We note how all four algorithms are unstable in similar regions of the plot (upper island as well as different parts of the shape in the lower left). In addition, we observe that the clustering methods, except Leiden, are more stable as the resolution value increases.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_clustering_difference_facet.html">plot_clustering_difference_facet</a></span><span class="op">(</span><span class="va">clustering_diff_obj</span>, <span class="va">cuomo</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">umap</span><span class="op">@</span><span class="va">cell.embeddings</span><span class="op">)</span></code></pre></div>
<p><img src="stability-based-parameter-assessment_files/figure-html/diff_facet-1.png" width="768"></p>
<p>As previously suggested, the resolution parameter affects the number of clusters obtained. The <code>get_resolution_importance</code> evaluates the effect of a range of values for the resolution parameter on the stability of the output. Besides the resolution, the user can input multiple values of the following parameters: number of neighbours, the graph type and the clustering method. The function will return a list associated with each parameter value combination. For each resolution value, the list will contain multiple sublists of partitions corresponding to specific number of clusters.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">resolution_gridsearch</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_resolution_importance.html">get_resolution_importance</a></span><span class="op">(</span>embedding <span class="op">=</span> <span class="va">cuomo</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">umap</span><span class="op">@</span><span class="va">cell.embeddings</span>,
                                                  resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0.5</span>, to <span class="op">=</span> <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,
                                                  n_neigh <span class="op">=</span> <span class="fl">30</span>,
                                                  n_repetitions <span class="op">=</span> <span class="va">n_repetitions</span>,
                                                  clustering_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,
                                                  graph_type <span class="op">=</span> <span class="fl">2</span>,
                                                  ecs_thresh <span class="op">=</span> <span class="fl">1</span>,
                                                  ncores <span class="op">=</span> <span class="va">n_cores</span><span class="op">)</span></code></pre></div>
<p>The resulting object can be visualized with <code>plot_k_resolution_corresp</code> as shown in the code below; we showcase the relationship between the number of clusters and the resolution value. This plot also provides information on suitable resolution values for predefined number of clusters. The colour gradient is proportional to the number of occurrences of the most frequent partition. It can also be used as proxy to describe the co-variation between (k, resolution). Lighter (higher) values indicate that little variation is observed, on changes of the random seed, between the number of clusters and the resolution value.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_k_resolution_corresp.html">plot_k_resolution_corresp</a></span><span class="op">(</span><span class="va">resolution_gridsearch</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"resolution - k correspondence with ecs threshold = 1"</span><span class="op">)</span></code></pre></div>
<p><img src="stability-based-parameter-assessment_files/figure-html/k_res_corr_1-1.png" width="768"></p>
<p>The following plot showcases the co-variation between the stability of the number of clusters and the number of different partitions resulting from changes on the seed or the resolution parameter. A high number of different partitions indicates a lower stability for a given number of clusters. The colour gradient is proportional to the frequency of the most common partition with a given number of clusters. We note that, even if we obtain a high number of different partitions, if the frequency of the most common one is close to 1, then the overall stability is high. Observing a high number of partitions, each with low frequency, indicates high instability.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_k_n_partitions.html">plot_k_n_partitions</a></span><span class="op">(</span><span class="va">resolution_gridsearch</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"k - # partitions correspondence with ecs threshold = 1"</span><span class="op">)</span></code></pre></div>
<p><img src="stability-based-parameter-assessment_files/figure-html/k_n_part_1-1.png" width="768"></p>
<p>By default, the <code>get_resolution_importance</code> function (as well as the previous functions part of the assessment pipeline) relies on the merging of identical partitions. Two partitions are identical if their contingency table form a square diagonal matrix. Merging is also applied to reduce the object size and to minimize the runtime for calculating the Element-Centric Consistency score. In practice, we also noticed cases with minor differences between two partitions (only 5 cells change their cluster labels). For the merging of identical partitions, these would be classed as different, although, in practice, the biological interpretation is likely to be the same. To address this we included a similarity threshold for additional flexibility i.e.: two partitions are considered identical and will be merged if the Element-Centric Similarity Score between them is above a user-defined ECS threshold.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">resolution_gridsearch_thresh_99</span> <span class="op">=</span> <span class="fu"><a href="../reference/merge_partitions.html">merge_partitions</a></span><span class="op">(</span><span class="va">resolution_gridsearch</span>,
                                                   ecs_thresh <span class="op">=</span> <span class="fl">0.99</span>,
                                                   ncores <span class="op">=</span> <span class="va">n_cores</span><span class="op">)</span></code></pre></div>
<p>Lowering the ECS threshold leads, in general, to higher number of occurences for the most frequent partition.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_k_resolution_corresp.html">plot_k_resolution_corresp</a></span><span class="op">(</span><span class="va">resolution_gridsearch_thresh_99</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"resolution - k correspondence with ecs threshold = 0.99"</span><span class="op">)</span></code></pre></div>
<p><img src="stability-based-parameter-assessment_files/figure-html/k_res_corr_99-1.png" width="768"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_k_n_partitions.html">plot_k_n_partitions</a></span><span class="op">(</span><span class="va">resolution_gridsearch_thresh_99</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"k - # partitions correspondence with ecs threshold = 0.99"</span><span class="op">)</span></code></pre></div>
<p><img src="stability-based-parameter-assessment_files/figure-html/k_n_part_99-1.png" width="768"></p>
<p>We conclude by summarizing the runtime for executing each step of the assessment pipeline. The results observed on other systems may vary; the runtime is also affected by the size of the date (i.e number of cells and features), the data itself and by actual values for the various parameters.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Feature stability methods runtime:"</span>,
      <span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">stop_time_feature_stability</span> <span class="op">-</span> <span class="va">start_time_feature_stability</span>,
                        units <span class="op">=</span> <span class="st">"mins"</span><span class="op">)</span><span class="op">)</span>, <span class="st">"minutes"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Feature stability methods runtime: 20.35029 minutes"</span>
<span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"NN - # connected components methods runtime:"</span>,
      <span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">stop_time_nn_conn</span> <span class="op">-</span> <span class="va">start_time_nn_conn</span>,
                        units <span class="op">=</span> <span class="st">"mins"</span><span class="op">)</span><span class="op">)</span>, <span class="st">"minutes"</span><span class="op">)</span>
<span class="co">#&gt; [1] "NN - # connected components methods runtime: 7.895996 minutes"</span>
<span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"NN importance methods runtime:"</span>,
      <span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">stop_time_nn_importance</span> <span class="op">-</span> <span class="va">start_time_nn_importance</span>,
                        units <span class="op">=</span> <span class="st">"mins"</span><span class="op">)</span><span class="op">)</span>, <span class="st">"minutes"</span><span class="op">)</span>
<span class="co">#&gt; [1] "NN importance methods runtime: 36.98898 minutes"</span>
<span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Clustering importance methods runtime:"</span>,
      <span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">stop_time_clustering_method</span> <span class="op">-</span> <span class="va">start_time_clustering_method</span>,
                        units <span class="op">=</span> <span class="st">"mins"</span><span class="op">)</span><span class="op">)</span>, <span class="st">"minutes"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Clustering importance methods runtime: 10.61586 minutes"</span>
<span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Resolution gridsearch methods runtime:"</span>,
      <span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">stop_time_resolution</span> <span class="op">-</span> <span class="va">start_time_resolution</span>,
                        units <span class="op">=</span> <span class="st">"mins"</span><span class="op">)</span><span class="op">)</span>, <span class="st">"minutes"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Resolution gridsearch methods runtime: 2.512017 minutes"</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h1>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Linux Mint 20.1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=ro_RO.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=ro_RO.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=ro_RO.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] patchwork_1.1.1    ClustAssess_0.3.0  ggplot2_3.3.5      Matrix_1.4-0      </span>
<span class="co">#&gt; [5] readr_2.1.0        SeuratObject_4.0.4 Seurat_4.0.5      </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] systemfonts_1.0.3     plyr_1.8.6            igraph_1.2.10        </span>
<span class="co">#&gt;   [4] lazyeval_0.2.2        splines_4.1.2         listenv_0.8.0        </span>
<span class="co">#&gt;   [7] scattermore_0.7       digest_0.6.29         foreach_1.5.1        </span>
<span class="co">#&gt;  [10] htmltools_0.5.2       fansi_0.5.0           magrittr_2.0.1       </span>
<span class="co">#&gt;  [13] memoise_2.0.1         tensor_1.5            cluster_2.1.2        </span>
<span class="co">#&gt;  [16] ROCR_1.0-11           tzdb_0.2.0            globals_0.14.0       </span>
<span class="co">#&gt;  [19] matrixStats_0.61.0    vroom_1.5.6           pkgdown_2.0.1        </span>
<span class="co">#&gt;  [22] spatstat.sparse_2.0-0 prettyunits_1.1.1     colorspace_2.0-2     </span>
<span class="co">#&gt;  [25] rappdirs_0.3.3        ggrepel_0.9.1         textshaping_0.3.6    </span>
<span class="co">#&gt;  [28] xfun_0.28             dplyr_1.0.7           crayon_1.4.2         </span>
<span class="co">#&gt;  [31] jsonlite_1.7.2        spatstat.data_2.1-0   survival_3.2-13      </span>
<span class="co">#&gt;  [34] zoo_1.8-9             iterators_1.0.13      glue_1.6.0           </span>
<span class="co">#&gt;  [37] polyclip_1.10-0       gtable_0.3.0          leiden_0.3.9         </span>
<span class="co">#&gt;  [40] future.apply_1.8.1    abind_1.4-5           scales_1.1.1         </span>
<span class="co">#&gt;  [43] DBI_1.1.1             miniUI_0.1.1.1        Rcpp_1.0.7           </span>
<span class="co">#&gt;  [46] progress_1.2.2        viridisLite_0.4.0     xtable_1.8-4         </span>
<span class="co">#&gt;  [49] reticulate_1.22       spatstat.core_2.3-2   bit_4.0.4            </span>
<span class="co">#&gt;  [52] htmlwidgets_1.5.4     httr_1.4.2            RColorBrewer_1.1-2   </span>
<span class="co">#&gt;  [55] ellipsis_0.3.2        ica_1.0-2             pkgconfig_2.0.3      </span>
<span class="co">#&gt;  [58] farver_2.1.0          sass_0.4.0            uwot_0.1.11          </span>
<span class="co">#&gt;  [61] deldir_1.0-6          here_1.0.1            utf8_1.2.2           </span>
<span class="co">#&gt;  [64] tidyselect_1.1.1      labeling_0.4.2        rlang_0.4.12         </span>
<span class="co">#&gt;  [67] reshape2_1.4.4        later_1.3.0           munsell_0.5.0        </span>
<span class="co">#&gt;  [70] tools_4.1.2           cachem_1.0.6          cli_3.1.0            </span>
<span class="co">#&gt;  [73] generics_0.1.1        ggridges_0.5.3        evaluate_0.14        </span>
<span class="co">#&gt;  [76] stringr_1.4.0         fastmap_1.1.0         yaml_2.2.1           </span>
<span class="co">#&gt;  [79] ragg_1.2.1            goftest_1.2-3         knitr_1.36           </span>
<span class="co">#&gt;  [82] bit64_4.0.5           fs_1.5.0              fitdistrplus_1.1-6   </span>
<span class="co">#&gt;  [85] purrr_0.3.4           RANN_2.6.1            pbapply_1.5-0        </span>
<span class="co">#&gt;  [88] future_1.23.0         nlme_3.1-153          mime_0.12            </span>
<span class="co">#&gt;  [91] compiler_4.1.2        rstudioapi_0.13       plotly_4.10.0        </span>
<span class="co">#&gt;  [94] curl_4.3.2            png_0.1-7             spatstat.utils_2.2-0 </span>
<span class="co">#&gt;  [97] tibble_3.1.6          bslib_0.3.1           stringi_1.7.6        </span>
<span class="co">#&gt; [100] highr_0.9             desc_1.4.0            RSpectra_0.16-0      </span>
<span class="co">#&gt; [103] lattice_0.20-45       vctrs_0.3.8           pillar_1.6.4         </span>
<span class="co">#&gt; [106] lifecycle_1.0.1       spatstat.geom_2.3-0   lmtest_0.9-39        </span>
<span class="co">#&gt; [109] jquerylib_0.1.4       RcppAnnoy_0.0.19      data.table_1.14.2    </span>
<span class="co">#&gt; [112] cowplot_1.1.1         irlba_2.3.5           httpuv_1.6.3         </span>
<span class="co">#&gt; [115] R6_2.5.1              promises_1.2.0.1      KernSmooth_2.23-20   </span>
<span class="co">#&gt; [118] gridExtra_2.3         parallelly_1.29.0     codetools_0.2-18     </span>
<span class="co">#&gt; [121] MASS_7.3-54           assertthat_0.2.1      rprojroot_2.0.2      </span>
<span class="co">#&gt; [124] withr_2.4.3           sctransform_0.3.2     mgcv_1.8-38          </span>
<span class="co">#&gt; [127] parallel_4.1.2        hms_1.1.1             grid_4.1.2           </span>
<span class="co">#&gt; [130] rpart_4.1-15          tidyr_1.1.4           rmarkdown_2.11       </span>
<span class="co">#&gt; [133] Rtsne_0.15            shiny_1.7.1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Arash Shahsavari, Andi Munteanu, Irina Mohorianu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
